pkg/agent/openflow/client.go

```diff
@@ -274,7 +275,26 @@ func (c *client) Initialize(roundNum uint64) (<-chan struct{}, error) {
 	// Ignore first notification, it is not a "reconnection".
 	<-connCh
 
-	c.cookieAllocator = cookie.NewAllocator(roundNum)
+	c.cookieAllocator = cookie.NewAllocator(roundInfo.RoundNum)
+
+	deleteStaleFlowsAfterConvergence := func() {
+		// Delete stale flows from previous round. We need to wait long enough to ensure
+		// that all the flow which are still required have received an updated cookie (with
+		// the new round number), otherwise we would disrupt the dataplane. Unfortunately,
+		// the time required for convergence is large and there is no simple way to
+		// determine when is a right time to perform the cleanup task.
+		// Additionally, with the current code, if the agent is restarted before having a
+		// chance to delete stale flows, these flows will never be deleted.
+		time.Sleep(10 * time.Second)
```

Comment:
Could we use a Channel or WaitGroup to make this deterministic? For example, this goroutine blocks on the Channel or WaitGroup, which will be unblocked by AgentInitializer, CNIServer, NodeRouteController, and NetworkPolicyController after they reconcile their rules. I think the first 3 components' time to send the signal is clear, it's not clear for NetworkPolicyController until its watching can converted to listing+versioned watching.
But we can have the timeout mechanism for now and enhance it later.

Reply:
Thanks @tnqn. I am definitely interested in improving this and I was considering something like the solution you describe.

However, even for `NodeRouteController`, I don't believe it is very straightforward. The current workflow is:
1. wait for caches to sync
2. call "reconcile", which only removes stale routes & stale tunnels
3. start worker goroutines, which will replay flows, etc.

So it's not obvious to me that there is a good way to determine when to remove stale flows. Of course, one solution is top modify "reconcile" to go over all the Nodes and replay the flows. When I wrote "reconcile", I avoided doing that because I want to let the workers do it "normally" and in-parallel.

From your comment it seems that `NetworkPolicyController` is even trickier.

Reply:
I will think about how to determine NetworkPolicyController has done an initial sync. I'm fine with using the timeout mechanism first.